ocaml/database/db_cache_types.ml

```diff
@@ -350,10 +350,10 @@ let remove_from_set key t =
   Schema.Value.Set (List.filter (fun x -> x <> key) t)
 
 exception Duplicate
-let add_to_map key value t =
+let add_to_map idempotent key value t =
   let t = Schema.Value.Unsafe_cast.pairs t in
-  if List.mem key (List.map fst t) then raise Duplicate;
-  Schema.Value.Pairs ((key, value) :: t)
+  if List.mem_assoc key t && (not idempotent || List.assoc key t <> value) then raise Duplicate;
```

Comment:
You are right - I now realise that you do `List.mem` to avoid a `Not_found` exception. This can stay as it is. But I think the best performance would have this:

```
try if not idempotent && List.assoc key t <> value then raise Duplicate 
with Not_found -> ()
```